### ISSUE TITLE
Using elasticsearch::user does not create a functioning user with xpack

### ISSUE THREAD

**Issue by hboetes** [elastic/puppet-elasticsearch#871]
Using a puppet recipe like below I can see `/etc/elasticsearch/elkel01/x-pack/users` `/etc/elasticsearch/elkel01/x-pack/users_roles` being edited with a proper password hash and role description and ES is restarted.

```
  elasticsearch::user {                                                                                                                                                                        
    'curator':                                                                                                                                                                            
      password => 'curator_password',                                                                                                                                                          
      roles    => ['superuser'];
  }                                                                                                                                                      
```

But those users don't show up in kibana and by using `curl` or `curator` I can see they don't have a real account. Not until I manually add them in the kibana interface.

I'm using the latest ES puppet module from git.

What's going wrong here?

**Comment by tylerjl** [comment]
This module manages [X-Pack's file realm](https://www.elastic.co/guide/en/x-pack/current/file-realm.html), which adds users via files and not over the [native user API](https://www.elastic.co/guide/en/x-pack/current/native-realm.html). I _think_ this is the underlying cause for your issue, as I don't think those file-based users are exposed over Kibana. Since both realms are supported by X-Pack, I believe you should still be able to use the API with the file-based user Puppet created.

**Comment by hboetes** [comment]
Indeed, it is highly frustrating to have two methods of local authentication that do not communicate with each other. I'm glad you agree it should work.

**Comment by hboetes** [comment]
You're right, I had to enable the file realm. I figured out the rest meanwhile with a lot of guessing, searching on google, working through error messages and more guessing. The documentation -- not just for this module, but also for elasticsearch in general -- is highly inadequate. The yaml syntax in no way reflects the api syntax or the webinterface. They even use completely different statements to achieve the same. And the native realm has no idea about the file realm.

This is what I came up with.
```
  elasticsearch::role {                                                                        
    'curator':                                                                                 
      privileges => {                                                                          
        'cluster' => ['monitor'],                                                              
        'indices' => [{                                                                        
          'names'      => ['*'],                                                               
          'privileges' => ['delete', 'index', 'view_index_metadata', 'monitor'],               
          }]                                                                                   
      };                                                                                       
    'logstash_writer':                                                                         
      privileges => {                                                                          
        'cluster' => ['monitor', 'manage_index_templates'],                                    
        'indices' => [{                                                                        
          'names'      => ['filebeat-*'],                                                      
          'privileges' => ['write', 'delete', 'create_index', 'index'],                        
          }]                                                                                   
      };                                                                                       
  }                                                                                            
```

**Comment by tylerjl** [comment]
@hboetes is there a particular section in this module's documentation that could be amended that would help clarify things?

Also, should we open an feature issue to support the native realm in addition to the file realm, or is the file-based realm sufficient?

**Comment by hboetes** [comment]
@tylerjl , thanks for asking, I wouldn't expect this module to contain a table with all the possible permissions of a roles.yml -- that would be out of scope, I'd expect it to be over here: https://www.elastic.co/guide/en/x-pack/current/defining-roles.html

But I found all the possible options in the curator error output which I c&p'ed as a comment in my code:

```
  # Error: TransportError(400, u'illegal_argument_exception',                                                                                                                                  
  # u'unknown cluster privilege [[monitor/settings/get]]. a privilege                                                                                                                          
  # must be either one of the predefined fixed cluster privileges                                                                                                                              
  # [all=[all],monitor_ml=[monitor_ml],manage_ingest_pipelines=[manage_ingest_pipelines],monitor_watcher=[monitor_watcher],manage_index_templates=[manage_index_templates],monitor=[monitor],n\
one=[none],manage_watcher=[manage_watcher],manage=[manage],manage_ml=[manage_ml],manage_pipeline=[manage_pipeline],transport_client=[transport_client],manage_security=[manage_security]]      
  # or a pattern over one of the available cluster actions') Error:                                                                                                                            
  # TransportError(400, u'illegal_argument_exception', u'unknown index                                                                                                                         
  # privilege [admin/get]. a privilege must be either one of the                                                                                                                               
  # predefined fixed indices privileges                                                                                                                                                        
  # [all=[all],read=[read],create_index=[create_index],read_cross_cluster=[read_cross_cluster],index=[index],monitor=[monitor],none=[none],delete=[delete],manage=[manage],delete_index=[delet\
e_index],view_index_metadata=[view_index_metadata],create=[create],write=[write]]                                                                                                              
  # or a pattern over one of the available index actions')                                                                                                                                     
```

And I found this example online: https://discuss.elastic.co/t/curator-delete-indices-shield-permissions/29354/4
Which does not match at all with the contents of a roles.yml

**Comment by tylerjl** [comment]
Thanks for the additional information @hboetes, hopefully it'll be useful for other users down the road 👍
